staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go

```diff
@@ -220,36 +222,66 @@ func (w *deferredResponseWriter) Write(p []byte) (n int, err error) {
 				attribute.Bool("firstWrite", firstWrite))
 		}
 	}()
+
+	switch {
+	case w.contentEncoding == "gzip":
+		return w.handleGzipEncoding(p)
+	default:
+		return w.handleDefaultEncoding(p)
+	}
+}
+
+func (w *deferredResponseWriter) handleDefaultEncoding(p []byte) (int, error) {
 	if w.hasWritten {
 		return w.w.Write(p)
 	}
 	w.hasWritten = true
 
 	hw := w.hw
 	header := hw.Header()
-	switch {
-	case w.contentEncoding == "gzip" && len(p) > defaultGzipThresholdBytes:
+	w.w = hw
+	header.Set("Content-Type", w.mediaType)
+	hw.WriteHeader(w.statusCode)
+	return w.w.Write(p)
+}
+
+func (w *deferredResponseWriter) handleGzipEncoding(p []byte) (int, error) {
+	if !w.isGzipped && len(w.gzipBuf)+len(p) > defaultGzipThresholdBytes {
+		w.isGzipped = true
+		w.gzipBuf = append(w.gzipBuf, p...)
+
+		header := w.hw.Header()
 		header.Set("Content-Encoding", "gzip")
 		header.Add("Vary", "Accept-Encoding")
+		header.Set("Content-Type", w.mediaType)
+		w.hw.WriteHeader(w.statusCode)
 
 		gw := gzipPool.Get().(*gzip.Writer)
-		gw.Reset(hw)
-
+		gw.Reset(w.hw)
 		w.w = gw
-	default:
-		w.w = hw
 	}
-
-	header.Set("Content-Type", w.mediaType)
-	hw.WriteHeader(w.statusCode)
-	return w.w.Write(p)
+	w.hasWritten = true
```

Comment:
Don't think this is correct place to set `hasWritten`. I expect it should only be set when we wrote to the `w.w` either directly or through gzip. One of reason to track `hasWritten` is to make sure we set headers correctly. 

I expect both `handleDefaultEncoding` and `handleGzipEncoding` should start from
```
	if w.hasWritten {
		return w.w.Write(p)
	}
```

So we might even would want to move it outside those functions.

Reply:
Outdated, let me close it